homeworks/13/instagram-api/instagram-api/src/main/java/elsys/instagram/api/PostServlet.java

```diff
@@ -0,0 +1,179 @@
+package elsys.instagram.api;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+@WebServlet("/posts/*")
+public class PostServlet extends HttpServlet { 
+    
+    private Gson gson;
+
+    @Override
+    public void init() throws ServletException {
+        gson = new Gson();
+        super.init();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String path = req.getServletContext().getContextPath();
+        PrintWriter out = resp.getWriter();
+        String username = req.getParameter("username");
+        JsonObject jsObj = new JsonObject(); 
+
+        if (path.matches("posts$")) {
+            for (Post post: Post.getPosts().values()) {
+                if (post.getUsername() == username) {
+                    jsObj.addProperty("post_id", post.getPostId());
+                }
+            }
+        } else if (path.matches("posts/(0|[0-9]+)/comments")) {
+            // check if posts exists
+            String[] ids = path.split("[0-9]+", 1);
+            int postId = Integer.parseInt(ids[0]);
+            
+            try {
+                Post post = Post.getPost(postId);
+                // get comments
+                for (Comment comment: post.getComments().values()) {
+                    jsObj.addProperty(Integer.toString(comment.getCommentId()), comment.getComment());
+                }
+            } catch (InvalidPostIdException e) {
+                resp.sendError(HttpServletResponse.SC_FORBIDDEN, e.getMessage());
+                return;
+            }
+        } else {
+            resp.sendError(HttpServletResponse.SC_NOT_FOUND);
+        }
+        
+        resp.setContentType("application/json");
+        out.print(jsObj);
+        out.flush();
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String path = req.getServletContext().getContextPath();
+        PrintWriter out = resp.getWriter();
+        User user;
+
+        try {
+            user = User.getUser(req.getParameter("username"));
```

Comment:
The way this logic is implemented allows anyone who knows the username of a certain user to post on their behalf. `HttpSessions` solve this problem with cookies.